src/reformulate.jl

```diff
@@ -1,82 +1,387 @@
+function _reformulate_logical_constraints(model::JuMP.Model)
+    for (_, lcon) in _logical_constraints(model)
+        _reformulate_logical_constraint(model, lcon.constraint.func, lcon.constraint.set)
     end
 end
-function reformulate_constraint(constr::AbstractArray{<:ConstraintRef}, bin_var, reformulation, param, i, j = missing)
-    for k in eachindex(constr)
-        reformulate_constraint(constr[k], bin_var, reformulation, param, i, j, k)
+
+function _reformulate_logical_constraint(model::JuMP.Model, lvec::Vector{LogicalVariableRef}, set::Union{MOIAtMost, MOIAtLeast, MOIExactly})
+    return _reformulate_selector(model, set, set.value, lvec)
+end
+
+function _reformulate_logical_constraint(model::JuMP.Model, lexpr::_LogicalExpr, ::_MOI.EqualTo{Bool})
+    return _reformulate_proposition(model, lexpr)
+end
+
+function _reformulate_selector(model::JuMP.Model, ::MOIAtLeast, val::Number, lvrefs::Vector{LogicalVariableRef})
+    bvrefs = _indicator_to_binary_ref.(lvrefs)
+    return JuMP.add_constraint(model,
+        JuMP.build_constraint(error, JuMP.@expression(model, sum(bvrefs)), _MOI.GreaterThan(val))
+    )
+end
+function _reformulate_selector(model::JuMP.Model, ::MOIAtMost, val::Number, lvrefs::Vector{LogicalVariableRef})
+    bvrefs = _indicator_to_binary_ref.(lvrefs)
+    return JuMP.add_constraint(model,
+        JuMP.build_constraint(error, JuMP.@expression(model, sum(bvrefs)), _MOI.LessThan(val))
+    )
+end
+function _reformulate_selector(model::JuMP.Model, ::MOIExactly, val::Number, lvrefs::Vector{LogicalVariableRef})
+    bvrefs = _indicator_to_binary_ref.(lvrefs)
+    return JuMP.add_constraint(model,
+        JuMP.build_constraint(error, JuMP.@expression(model, sum(bvrefs)), _MOI.EqualTo(val))
+    )
+end
+function _reformulate_selector(model::JuMP.Model, ::MOIAtLeast, lvref::LogicalVariableRef, lvrefs::Vector{LogicalVariableRef})
+    bvref = _indicator_to_binary_ref(lvref)
+    bvrefs = Vector{Any}(_indicator_to_binary_ref.(lvrefs))
+    return JuMP.add_constraint(model,
+        build_constraint(error, JuMP.NonlinearExpr(:-, Any[JuMP.NonlinearExpr(:+, bvrefs), bvref]), _MOI.GreaterThan(0))
+    )
+end
+function _reformulate_selector(model::JuMP.Model, ::MOIAtMost, lvref::LogicalVariableRef, lvrefs::Vector{LogicalVariableRef})
+    bvref = _indicator_to_binary_ref(lvref)
+    bvrefs = Vector{Any}(_indicator_to_binary_ref.(lvrefs))
+    return JuMP.add_constraint(model,
+        build_constraint(error, JuMP.NonlinearExpr(:-, Any[JuMP.NonlinearExpr(:+, bvrefs), bvref]), _MOI.LessThan(0))
+    )
+end
+function _reformulate_selector(model::JuMP.Model, ::MOIExactly, lvref::LogicalVariableRef, lvrefs::Vector{LogicalVariableRef})
+    bvref = _indicator_to_binary_ref(lvref)
+    bvrefs = Vector{Any}(_indicator_to_binary_ref.(lvrefs))
+    return JuMP.add_constraint(model,
+        build_constraint(error, JuMP.NonlinearExpr(:-, Any[JuMP.NonlinearExpr(:+, bvrefs), bvref]), _MOI.EqualTo(0))
+    )
+end
```

Comment:
Are these methods really needed? Having a logic variable for `n` doesn't seem to have any practical use.

Reply:
It does for Nested relationships where the enforcement of the logical constraint depends on the parent.

Reply:
Ok, we'll want to refactor these to create affine constraints then instead of nonlinear ones.

Reply:
done